cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/MetricsMessageAggregator.java

```diff
@@ -0,0 +1,297 @@
+ * </ol>
+ */
+class MetricsMessageAggregator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      MetricsMessageAggregator.class);
+
+  private final Map<String, String> cachemap;
+  private final ResettableByteArrayInputStream decodingInputStream;
+  private final StringCachingDecoder decoder;
+  private final DatumReader<MetricValues> reader;
+  private final Schema schema;
+  private final ByteArrayOutputStream encodeOutputStream;
+  private final BinaryEncoder encoder;
+  private final DatumWriter<MetricValues> recordWriter;
+  private final long aggregationWindowSeconds;
+
+  MetricsMessageAggregator(SchemaGenerator schemaGenerator,
+      DatumWriterFactory writerFactory, DatumReaderFactory readerFactory,
+      long aggregationWindowSeconds) {
+    TypeToken<MetricValues> metricValueType = TypeToken.of(MetricValues.class);
+    try {
+      this.schema = schemaGenerator.generate(metricValueType.getType());
+    } catch (UnsupportedTypeException e) {
+      throw new RuntimeException("Failed to generate schema for MetricValues",
+          e);
+    }
+    this.reader = readerFactory.create(metricValueType, schema);
+    this.recordWriter = writerFactory.create(metricValueType, schema);
+    this.cachemap = new HashMap<>();
+    this.decodingInputStream = new ResettableByteArrayInputStream();
+    this.decoder = new StringCachingDecoder(
+        new BinaryDecoder(decodingInputStream), cachemap);
+    this.encodeOutputStream = new ByteArrayOutputStream(1024);
+    this.encoder = new BinaryEncoder(encodeOutputStream);
+    this.aggregationWindowSeconds = Math.max(1, aggregationWindowSeconds);
+  }
+
+  public Iterator<Message> aggregate(Iterator<Message> input) {
+    if (!input.hasNext()) {
+      return input;
+    }
+    cachemap.clear();
+
+    Map<MetricsMapKey, MetricsAggregator> aggregated = new HashMap<>();
+    String[] messageId = new String[1];
+    long currentTimestamp =
+        System.currentTimeMillis() / TimeUnit.SECONDS.toMillis(1);
+    // MetricValues will be aggregated across windows of size
+    // "aggregationWindowSeconds". We align the current time to the end of the
+    // last widow. This is done since in most cases, it is expected that all
+    // the MetricValues were emitted very recently. The offset ensures
+    // the last window completely lies in the past and maximizes the chance
+    // that all metrics lie within it.
+    long offset = (currentTimestamp % aggregationWindowSeconds) + 1;
+
+    input.forEachRemaining(message -> {
+      // Decode the metrics.
+      // Keep track of the last message ID as TopicRelayers need it to fetch
+      // messages in the next iteration.
+      messageId[0] = message.getId();
+      decodingInputStream.reset(message.getPayload());
+      MetricValues values;
+      try {
+        values = reader.read(decoder, schema);
+      } catch (IOException e) {
+        LOG.warn("Failed to decode metric message.", e);
```

Comment:
Do we need to stop in case of IOException? Can't we continue without aggregation if possible?

Reply:
Made changes to return partially aggregated messages on failure. 
Now the metrics that have been aggregated before the failure + the messages that are not aggregated are sent back. Added a test for the same.